### PR TITLE
feat: add a reverse db walker

### LINKED ISSUE #837: Add reverse db walker
### Describe the feature

Similar to the current walker but it will use `prev` function to iterate in reverse.

There are a few places that would benefit from it.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by CryptoTurboFish**
Closes #837

### CODE REVIEW

**Comment by rakita** [comment]
This looks good! Thank you

Can you add functions to the cursors, that would make this whole: https://github.com/paradigmxyz/reth/blob/main/crates/storage/db/src/abstraction/cursor.rs

File: crates/storage/db/src/abstraction/cursor.rs [hunk 1]
```diff
@@ -122,6 +122,32 @@ impl<'cursor, 'tx, T: Table, CURSOR: DbCursorRO<'tx, T>> std::iter::Iterator
     }
 }
 
+/// Provides a reverse iterator to `Cursor` when handling `Table`.
+/// Also check [`Walker`]
+pub struct RWalker<'cursor, 'tx, T: Table, CURSOR: DbCursorRO<'tx, T>> {
```
Comment (gakonst) [inline crates/storage/db/src/abstraction/cursor.rs:127]:
```suggestion
pub struct ReverseWalker<'cursor, 'tx, T: Table, CURSOR: DbCursorRO<'tx, T>> {
```

File: crates/storage/db/src/abstraction/cursor.rs [hunk 2]
```diff
@@ -122,6 +122,32 @@ impl<'cursor, 'tx, T: Table, CURSOR: DbCursorRO<'tx, T>> std::iter::Iterator
...
+    /// Cursor to be used to walk through the table.
+    pub cursor: &'cursor mut CURSOR,
+    /// `(key, value)` where to start the walk.
+    pub start: IterPairResult<T>,
+    /// Phantom data for 'tx. As it is only used for `DbCursorRO`.
+    pub _tx_phantom: PhantomData<&'tx T>,
+}
+
+impl<'cursor, 'tx, T: Table, CURSOR: DbCursorRO<'tx, T>> std::iter::Iterator
```
Comment (gakonst) [inline crates/storage/db/src/abstraction/cursor.rs:136]: Let's add a method called rev() on walker which converts it to a ReverseWalker and vice versa?

File: crates/storage/db/src/implementation/mdbx/mod.rs [hunk 3]
```diff
@@ -213,6 +213,56 @@ mod tests {
...
+            .try_for_each(|key| tx.put::<CanonicalHeaders>(key, H256::zero()))
+            .expect(ERROR_PUT);
+        tx.commit().expect(ERROR_COMMIT);
+
+        let missing_key = 2;
+        let tx = db.tx().expect(ERROR_INIT_TX);
+        let mut cursor = tx.cursor_read::<CanonicalHeaders>().unwrap();
+
+        let mut walker =
+            Walker { cursor: &mut cursor, start: None, _tx_phantom: std::marker::PhantomData };
```
Comment (gakonst) [inline crates/storage/db/src/implementation/mdbx/mod.rs:233]: Don't we have a new method on the walker? If not let's add it

File: crates/storage/db/src/implementation/mdbx/mod.rs [hunk 4]
```diff
@@ -213,6 +213,56 @@ mod tests {
...
+
+        assert_eq!(walker.next(), Some(Ok((missing_key - 2, H256::zero()))));
+        assert_eq!(walker.next(), Some(Ok((missing_key - 1, H256::zero()))));
+        assert_eq!(walker.next(), Some(Ok((missing_key + 1, H256::zero()))));
```
Comment (gakonst) [inline crates/storage/db/src/implementation/mdbx/mod.rs:237]: These keys are a bit confusing, let's call them explicitly 0, 1, 3

File: crates/storage/db/src/implementation/mdbx/mod.rs [hunk 5]
```diff
@@ -213,6 +213,56 @@ mod tests {
...
+        tx.commit().expect(ERROR_COMMIT);
+
+        let missing_key = 2;
+        let tx = db.tx().expect(ERROR_INIT_TX);
+        let mut cursor = tx.cursor_read::<CanonicalHeaders>().unwrap();
+
+        let mut r_walker =
+            RWalker { cursor: &mut cursor, start: None, _tx_phantom: std::marker::PhantomData };
+
+        assert_eq!(r_walker.next(), Some(Ok((missing_key + 1, H256::zero()))));
```
Comment (gakonst) [inline crates/storage/db/src/implementation/mdbx/mod.rs:260]: Same comment as above re keys

**Review by gakonst — CHANGES_REQUESTED** [review]
Nits lgtm mostly pending Dragan comments

File: crates/storage/db/src/abstraction/cursor.rs (hunk 2 above)
Comment (CryptoTurboFish) [reply crates/storage/db/src/abstraction/cursor.rs:152]: like the idea^

File: crates/storage/db/src/implementation/mdbx/mod.rs (hunk 3 above)
Comment (CryptoTurboFish) [reply crates/storage/db/src/implementation/mdbx/mod.rs:233]: i don't think we have it, actually do we need to keep the fields public? Was checking around the usage seems we can make the fields private. 

And it makes sense to me if we just keep internal state of `Walker` private exposing only iteration method.

```
pub struct Walker<'cursor, 'tx, T: Table, CURSOR: DbCursorRO<'tx, T>> {
    /// Cursor to be used to walk through the table.
    pub cursor: &'cursor mut CURSOR,
    /// `(key, value)` where to start the walk.
    pub start: IterPairResult<T>,
    /// Phantom data for 'tx. As it is only used for `DbCursorRO`.
    pub _tx_phantom: PhantomData<&'tx T>,
}
```

File: crates/storage/db/src/abstraction/cursor.rs [hunk 6]
```diff
@@ -122,6 +122,66 @@ impl<'cursor, 'tx, T: Table, CURSOR: DbCursorRO<'tx, T>> std::iter::Iterator
     }
 }
 
+impl<'cursor, 'tx, T: Table, CURSOR: DbCursorRO<'tx, T>> Walker<'cursor, 'tx, T, CURSOR> {
+    /// construct Walker
+    pub fn new(
+        cursor: &'cursor mut CURSOR,
+        start: IterPairResult<T>,
+        _tx_phantom: PhantomData<&'tx T>,
```
Comment (gakonst) [inline crates/storage/db/src/abstraction/cursor.rs:130]: Can't you omit this param and have it be generated inside the function?

Comment (CryptoTurboFish) [reply crates/storage/db/src/abstraction/cursor.rs]: fixed, much cleaner

**Review by gakonst — APPROVED** [review]
Merging and we can iterate as needed

**Pull request merged**